infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/LogReplicationClientServerRouter.java

```diff
@@ -0,0 +1,605 @@
+import java.util.concurrent.atomic.AtomicLong;
+
+import static org.corfudb.protocols.CorfuProtocolCommon.getUuidMsg;
+import static org.corfudb.protocols.service.CorfuProtocolMessage.getDefaultProtocolVersionMsg;
+import static org.corfudb.protocols.service.CorfuProtocolMessage.getRequestMsg;
+import static org.corfudb.protocols.service.CorfuProtocolMessage.getResponseMsg;
+
+/**
+ * A Base Source Router, which interfaces between the Log replication Source components (FSMs, AckReader, etc) and the
+ * custom transport adapter
+ *
+ * This class is inherited by both LogReplicationSourceClientRouter and LogReplicationSourceServerRouter.
+ */
+@Slf4j
+public class LogReplicationClientServerRouter implements IClientServerRouter {
+
+    public static String REMOTE_LEADER = "REMOTE_LEADER";
+
+    /**
+     * A map of remoteCluster to {@link CompletableFuture} for each session, which is completed when a connection,
+     * including a successful handshake, completes and messages can be sent
+     * to the remote node.
+     */
+    @Getter
+    private volatile Map<LogReplicationSession, CompletableFuture<Void>> sessionToLeaderConnectionFuture;
+
+    /**
+     * A map of session to the current requestID for the session.
+     */
+    @Getter
+    @SuppressWarnings("checkstyle:abbreviation")
+    private final Map<LogReplicationSession,AtomicLong> sessionToRequestIdCounter;
+
+    /**
+     * Sync call response timeout (milliseconds).
+     */
+    @Getter
+    @Setter
+    private long timeoutResponse;
+
+    /**
+     * The outstanding requests on this router.
+     */
+    @Getter
+    private final Map<LogReplicationSession, Map<Long, CompletableFuture>> sessionToOutstandingRequests;
+
+    /**
+     * Runtime FSM, to insert connectivity events
+     */
+    @Getter
+    private final Map<LogReplicationSession, CorfuLogReplicationRuntime> sessionToRuntimeFSM;
+
+    /**
+     * Remote cluster's clusterDescriptor
+     */
+    @Getter
+    private final Map<LogReplicationSession, ClusterDescriptor> sessionToRemoteClusterDescriptor;
+
+    /**
+     * The replicationManager
+     */
+    private final CorfuReplicationManager replicationManager;
+
+    /**
+     * local cluster ID
+     */
+    private String localClusterId;
+
+    /**
+     * Client Transport adapter
+     */
+    @Setter
+    private IClientChannelAdapter clientChannelAdapter;
+
+    /**
+     * Server transport adapter
+     */
+    @Setter
+    @Getter
+    private IServerChannelAdapter serverChannelAdapter;
+
+    /**
+     * The handlers registered to this router.
+     */
+    @Getter
+    private final LogReplicationServer msgHandler;
+
+    /**
+     * Triggers leadership request and create bidirectional streams to remote endpoints
+     */
+    private final Map<LogReplicationSession, VerifyRemoteSourceLeader> sessionToSinkVerifyLeadership;
+
+    /**
+     * Session manager to check session details
+     */
+    private final SessionManager sessionManager;
```

Comment:
this is a circular dependency - sessionManager creates this router which then depends on it.  SessionManager can pass all the required params on creation itself.